File: patterns/06_dependency_injection/csharp/Officium.cs
// ============================================================
//  DEPENDENCY INJECTION — The Praetorian Quartermaster
//  Run: dotnet script Officium.cs
// ============================================================
using System; using System.Collections.Generic;

interface IMessenger { void Send(string to, string msg); }
interface ILogger    { void Log(string ev); }

record ScrollMessenger() : IMessenger { public void Send(string to, string msg) => Console.WriteLine($"  📜 Scroll to {to}: '{msg}'"); }
class MockMessenger : IMessenger {
    public List<(string, string)> Sent = new();
    public void Send(string to, string msg) { Sent.Add((to, msg)); Console.WriteLine($"  🧪 MOCK: to={to} msg='{msg}'"); }
}
record ConsoleLogger() : ILogger { public void Log(string ev) => Console.WriteLine($"  [LOG] {ev}"); }

class MilitaryCampaign {
    private readonly IMessenger _m; private readonly ILogger _l;
    public MilitaryCampaign(IMessenger m, ILogger l) { _m = m; _l = l; }
    public void LaunchAttack(string target) {
        _l.Log($"Campaign against {target}");
        _m.Send("Praetorian Guard", $"ADVANCE on {target}!");
        _m.Send("Supply Corps",     $"Provisions to {target}");
        _l.Log("Orders dispatched");
    }
}

Console.WriteLine("╔═══════════════════════════════════════════════╗");
Console.WriteLine("║   DEPENDENCY INJECTION — C#                   ║");
Console.WriteLine("╚═══════════════════════════════════════════════╝\n");
Console.WriteLine("── PRODUCTION ───────────────────────────────────");
new MilitaryCampaign(new ScrollMessenger(), new ConsoleLogger()).LaunchAttack("Gaul");
Console.WriteLine("\n── TESTING (mock) ───────────────────────────────");
var mock = new MockMessenger();
new MilitaryCampaign(mock, new ConsoleLogger()).LaunchAttack("Britannia");
Console.WriteLine($"  Mock captured: {mock.Sent.Count} messages ✓");
Console.WriteLine("\n\"Non ipse quaerit, accipit!\"");

File: patterns/06_dependency_injection/php/officium.php
<?php
/**
 * DEPENDENCY INJECTION — The Praetorian Quartermaster
 * Run: php officium.php
 */
declare(strict_types=1);

interface IMessenger { public function send(string $to, string $msg): void; }
interface ILogger    { public function log(string $event): void; }

class ScrollMessenger implements IMessenger {
    public function send(string $to, string $msg): void {
        echo "  📜 Scroll to {$to}: '{$msg}'\n";
    }
}
class MockMessenger implements IMessenger {
    public array $sent = [];
    public function send(string $to, string $msg): void {
        $this->sent[] = [$to, $msg];
        echo "  🧪 MOCK to={$to}: '{$msg}'\n";
    }
}
class ConsoleLogger implements ILogger {
    public function log(string $event): void { echo "  [LOG] {$event}\n"; }
}

class MilitaryCampaign {
    public function __construct(
        private IMessenger $messenger,
        private ILogger    $logger
    ) {}
    public function launchAttack(string $target): void {
        $this->logger->log("Campaign against {$target}");
        $this->messenger->send("Praetorian Guard", "ADVANCE on {$target}!");
        $this->messenger->send("Supply Corps",     "Provisions to {$target}");
        $this->logger->log("Orders dispatched");
    }
}

echo "╔═══════════════════════════════════════════════╗\n";
echo "║   DEPENDENCY INJECTION — PHP                  ║\n";
echo "║   The Praetorian Quartermaster                ║\n";
echo "╚═══════════════════════════════════════════════╝\n\n";

echo "── PRODUCTION ───────────────────────────────────\n";
(new MilitaryCampaign(new ScrollMessenger(), new ConsoleLogger()))->launchAttack("Gaul");

echo "\n── TESTING (mock) ───────────────────────────────\n";
$mock = new MockMessenger();
(new MilitaryCampaign($mock, new ConsoleLogger()))->launchAttack("Britannia");
echo "  Mock captured: " . count($mock->sent) . " messages ✓\n";
echo "\n\"Non ipse quaerit, accipit!\"\n";
